orttraining/orttraining/training_ops/cpu/controlflow/ort_tasks.h

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/common.h"
+#include "core/framework/ml_value.h"
+
+#include <mutex>
+#include <future>
+
+
+namespace onnxruntime {
+namespace contrib {
+
+typedef std::pair<Status, std::vector<OrtValue>> ForwardReturnType;
+typedef std::pair<bool, std::vector<OrtValue>> BackwardReturnType;
+
+class OrtTasks final {
+ public:
+  static OrtTasks& GetInstance() {
+    static OrtTasks* instance_ = new OrtTasks;
+    return *instance_;
+  }
+
+  void CreateBackgroundTask(int64_t run_id);
+  void RemoveTask(int64_t run_id);
+
+  void SetForwardOutputs(Status s, const std::vector<OrtValue>& forward_outputs);
+  ForwardReturnType WaitForForwardOutputs(int64_t run_id);
+  bool ForwardOutputsIsValid();
+
+  void SetBackwardInputs(int64_t run_id, const std::vector<OrtValue>& backward_inputs, bool terminate);
+  BackwardReturnType WaitForBackwardInputs();
+
+  void SetStatus(const Status& status);
+  Status WaitForStatus(int64_t run_id);
+  bool TaskIsCompleted(int64_t run_id);
+
+ private:
+  OrtTasks() = default;
+  ~OrtTasks() = default;
+  OrtTasks(const OrtTasks&) = delete;
+  OrtTasks& operator=(const OrtTasks&) = delete;
+
+  struct Task {
+    std::promise<ForwardReturnType> forward_output_promise_;
+    std::future<ForwardReturnType> forward_output_future_ = forward_output_promise_.get_future();
+
+    std::promise<BackwardReturnType> backward_input_promise_;
+    std::future<BackwardReturnType> backward_input_future_ = backward_input_promise_.get_future();
+
+    std::promise<Status> status_promise_;
+    std::future<Status> status_future_ = status_promise_.get_future();
+  };
+
+  std::hash<std::thread::id> hasher_;
+  mutable std::mutex mutex_;
```

Comment:
>mutable std::mutex mutex_; [](start = 2, length = 26)

since there is a mutable data member, does this mean some member functions should be const?

Reply:
right. 
ForwardOutputsIsValid and TaskIsCompleted can be const.